transform/coroutines.go

```diff
@@ -0,0 +1,934 @@
+	// tailCalls is the set of all tail calls in the asyncFunc.
+	calls, normalCalls, tailCalls map[llvm.Value]struct{}
+}
+
+// coroutineLoweringPass is a goroutine lowering pass which is used with the "coroutines" scheduler.
+type coroutineLoweringPass struct {
+	mod     llvm.Module
+	ctx     llvm.Context
+	builder llvm.Builder
+
+	// asyncFuncs is a map of all asyncFuncs.
+	// The map keys are function pointers.
+	asyncFuncs map[llvm.Value]*asyncFunc
+
+	// calls is a slice of all of the async calls in the module.
+	calls []llvm.Value
+
+	i8ptr llvm.Type
+
+	// memory management functions from the runtime
+	alloc, free llvm.Value
+
+	// coroutine intrinsics
+	start, pause, current                                   llvm.Value
+	setState, setRetPtr, getRetPtr, returnTo, returnCurrent llvm.Value
+	createTask                                              llvm.Value
+
+	// llvm.coro intrinsics
+	coroId, coroSize, coroBegin, coroSuspend, coroEnd, coroFree, coroSave llvm.Value
+}
+
+// findAsyncFuncs finds all asynchronous functions.
+// A function is considered asynchronous if it calls an asynchronous function or intrinsic.
+func (c *coroutineLoweringPass) findAsyncFuncs() {
+	asyncs := map[llvm.Value]*asyncFunc{}
+	calls := []llvm.Value{}
+
+	// Use a breadth-first search to find all async functions.
+	worklist := []llvm.Value{c.pause}
+	for len(worklist) > 0 {
+		// Pop a function off the worklist.
+		fn := worklist[len(worklist)-1]
+		worklist = worklist[:len(worklist)-1]
+
+		// Get task pointer argument.
+		task := fn.LastParam()
+		if fn != c.pause && (task.IsNil() || task.Name() != "parentHandle") {
+			panic("trying to make exported function async: " + fn.Name())
+		}
+
+		// Search all uses of the function while collecting callers.
+		callers := map[llvm.Value]struct{}{}
+		for use := fn.FirstUse(); !use.IsNil(); use = use.NextUse() {
+			user := use.User()
+			if user.IsACallInst().IsNil() {
+				// User is not a call instruction, so this is irrelevant.
+				continue
+			}
+			if user.CalledValue() != fn {
+				// Not the called value.
+				continue
+			}
+
+			// Add to calls list.
+			calls = append(calls, user)
+
+			// Get the caller.
+			caller := user.InstructionParent().Parent()
+
+			// Add as caller.
+			callers[caller] = struct{}{}
+
+			if _, ok := asyncs[caller]; ok {
+				// Already marked caller as async.
+				continue
+			}
+
+			// Mark the caller as async.
+			// Use nil as a temporary value. It will be replaced later.
+			asyncs[caller] = nil
+
+			// Put the caller on the worklist.
+			worklist = append(worklist, caller)
+		}
+
+		asyncs[fn] = &asyncFunc{
+			fn:      fn,
+			rawTask: task,
+			callers: callers,
+		}
+	}
+
+	c.asyncFuncs = asyncs
+	c.calls = calls
+}
+
+// ErrMissingIntrinsic is an error indicating that a required intrinsic was not found in the module.
+type ErrMissingIntrinsic struct {
+	Name string
+}
+
+func (err ErrMissingIntrinsic) Error() string {
+	return "missing intrinsic: " + err.Name
+}
```

Comment:
I think it would be sensible to move this to errors.go, what do you think @jaddr2line? Other passes may also want to emit this kind of error.